src/EventStore.Core/Services/RequestManager/CommitSource.cs

```diff
@@ -1,72 +1,34 @@
 ﻿using System;
 using System.Collections.Concurrent;
-using System.Collections.Generic;
 using System.Linq;
 using System.Threading;
+using System.Threading.Tasks;
 using EventStore.Core.Bus;
 using EventStore.Core.Messages;
 
 namespace EventStore.Core.Services.RequestManager {
 	public class CommitSource :
 	IHandle<ReplicationTrackingMessage.IndexedTo>,
 	IHandle<ReplicationTrackingMessage.ReplicatedTo> {
-		private readonly ConcurrentDictionary<long, List<Action>> _notifyReplicated = new ConcurrentDictionary<long, List<Action>>();
-		private readonly ConcurrentDictionary<long, List<Action>> _notifyIndexed = new ConcurrentDictionary<long, List<Action>>();
-		private long _replicatedPosition;
-		private long _indexedPosition;
+		private LogNotificationService _indexTracker = new LogNotificationService("IndexedTracker");
+		private LogNotificationService _replicatedTracker = new LogNotificationService("ReplicatedTracker");
 
-		public long ReplicationPosition => _replicatedPosition;
-		public long IndexedPosition => _indexedPosition;
-
-		public void Handle(ReplicationTrackingMessage.ReplicatedTo message) {
-			Interlocked.Exchange(ref _replicatedPosition, message.LogPosition);
-			Notify(_notifyReplicated, message.LogPosition);
+		public void Handle(ReplicationTrackingMessage.ReplicatedTo @event) {
+			_replicatedTracker.UpdateLogPosition(@event.LogPosition);
 		}
-
-		public void Handle(ReplicationTrackingMessage.IndexedTo message) {
-			Interlocked.Exchange(ref _indexedPosition, message.LogPosition);
-			Notify(_notifyIndexed, message.LogPosition);
+		public void Handle(ReplicationTrackingMessage.IndexedTo @event) {
+			_indexTracker.UpdateLogPosition(@event.LogPosition);
 		}
-
-		public void NotifyFor(long position, Action target, CommitLevel level = CommitLevel.Indexed) {
-			long currentPosition;
-			ConcurrentDictionary<long, List<Action>> notificationDictionary;
-			switch (level) {
-				case CommitLevel.Replicated:
-					currentPosition = Interlocked.Read(ref _replicatedPosition);
-					notificationDictionary = _notifyReplicated;
-					break;
-				case CommitLevel.Indexed:
-					currentPosition = Interlocked.Read(ref _indexedPosition);
-					notificationDictionary = _notifyIndexed;
-					break;
-				default:
-					throw new ArgumentOutOfRangeException(nameof(level), level, null);
-			}
-			if (currentPosition >= position) { target(); }
-			if (!notificationDictionary.TryGetValue(position, out var actionList)) {
-				actionList = new List<Action> { target };
-				notificationDictionary.TryAdd(position, actionList);
-			} else {
-				lock (actionList) {
-					actionList.Add(target);
-				}
-			}
+		public Task WaitForReplication(long position, CancellationToken token) {
+			return new Task(async () => {
```

Comment:
The returned Task is in the "created" state and won't never complete, i.e. `await` that task and you have a deadlock.

Why not
```c#
return _replicatedTracker.Waitfor(position);
```
?
The CancellationToken isn't passed in, but the same problem -- amongst the deadlock -- has the current state of PR too.
Ideally the CancellationToken should be passed down to `_replicatedTracker.Waitfor`.